packages/toc/src/utils/markdown.ts

```diff
@@ -0,0 +1,266 @@
+  for (let lineIdx = 0; lineIdx < lines.length; lineIdx++) {
+    let line = lines[lineIdx];
+
+    if (line === '') {
+      // Bail early
+      continue;
+    }
+
+    // Don't check for Markdown headings if in a code block:
+    if (line.startsWith('```')) {
+      isCodeBlock = !isCodeBlock;
+    }
+    if (isCodeBlock) {
+      continue;
+    }
+
+    const heading = parseHeading(line, lines[lineIdx + 1]); // append the next line to capture alternative style Markdown headings
+
+    if (heading) {
+      let level = heading.level;
+
+      if (level > 0 && level <= config.maximalDepth) {
+        const prefix = getPrefix(level, previousLevel, levels, config);
+        previousLevel = level;
+
+        headings.push({
+          text: heading.text,
+          prefix,
+          level,
+          line: lineIdx,
+          raw: heading.raw
+        });
+      }
+    }
+  }
+  return headings;
+}
+
+const MARKDOWN_MIME_TYPE = [
+  'text/x-ipythongfm',
+  'text/x-markdown',
+  'text/x-gfm',
+  'text/markdown'
+];
+
+/**
+ * Returns whether a MIME type corresponds to a Markdown flavor.
+ *
+ * @param mime - MIME type string
+ * @returns boolean indicating whether a provided MIME type corresponds to a Markdown flavor
+ *
+ * @example
+ * const bool = isMarkdown('text/markdown');
+ * // returns true
+ *
+ * @example
+ * const bool = isMarkdown('text/plain');
+ * // returns false
+ */
+export function isMarkdown(mime: string): boolean {
+  return MARKDOWN_MIME_TYPE.includes(mime);
+}
+
+/**
+ * Interface describing a parsed heading result.
+ *
+ * @private
+ */
+interface IHeader {
+  /**
+   * Heading text.
+   */
+  text: string;
+
+  /**
+   * Heading level.
+   */
+  level: number;
+
+  /**
+   * Raw string containing the heading
+   */
+  raw: string;
+}
+
+/**
+ * Parses a heading, if one exists, from a provided string.
+ *
+ * ## Notes
+ *
+ * -   Heading examples:
+ *
+ *     -   Markdown heading:
+ *
+ *         ```
+ *         # Foo
+ *         ```
+ *
+ *     -   Markdown heading (alternative style):
+ *
+ *         ```
+ *         Foo
+ *         ===
+ *         ```
+ *
+ *         ```
+ *         Foo
+ *         ---
+ *         ```
+ *
+ *     -   HTML heading:
+ *
+ *         ```
+ *         <h3>Foo</h3>
+ *         ```
+ *
+ * @private
+ * @param line - Line to parse
+ * @param nextLine - The line after the one to parse
+ * @returns heading info
+ *
+ * @example
+ * const out = parseHeading('### Foo\n');
+ * // returns {'text': 'Foo', 'level': 3}
+ *
+ * @example
+ * const out = parseHeading('Foo\n===\n');
+ * // returns {'text': 'Foo', 'level': 1}
+ *
+ * @example
+ * const out = parseHeading('<h4>Foo</h4>\n');
+ * // returns {'text': 'Foo', 'level': 4}
+ *
+ * @example
+ * const out = parseHeading('Foo');
+ * // returns null
+ */
+function parseHeading(line: string, nextLine?: string): IHeader | null {
+  // Case: Markdown heading
+  let match = line.match(/^([#]{1,6}) (.*)/);
+  if (match) {
+    if (!skipHeading.test(match[0])) {
```

Comment:
Looks like the CodeQL failure might be relevant: https://github.com/jupyterlab/jupyterlab/pull/12374/checks?check_run_id=6197070826

![image](https://user-images.githubusercontent.com/591645/165933629-e5848276-efcb-401f-97b7-4a3f340afdc9.png)

Reply:
I have no clue how to improve those regex as we are not looking for very specific matches but for some structures. There are three of them (from which only the first one is new):

```ts
/**
 * Ignore title with html tag with a class name equal to `jp-toc-ignore` or `tocSkip`
 */
const skipHeading = /<\w+\s(.*?\s)?class="(.*?\s)?(jp-toc-ignore|tocSkip)(\s.*?)?"(\s.*?)?>/;

// Case: HTML heading (WARNING: this is not particularly robust, as HTML headings can span multiple lines)
  match = line.match(/<h([1-6]).*>(.*)<\/h\1>/i);
  
// take special care to parse Markdown links into raw text
  return heading.replace(/\[(.+)\]\(.+\)/g, '$1');
```

Reply:
> from which only the first one is new

So maybe the CodeQL check caught it because the diff touched the file.

Reply:
Yes

Reply:
@fcollonval mind posting a link to the same regular expression already on `master`?

Just as reference so it's clear this was already there before the change.